File: algorithms/java/436_Read_N_Characters_Given_Read4_Call_Multiple_times.java
/**
 * The read4 API is defined in the parent class Reader4.
 *     int read4(char[] buf); 
 */
public class Solution extends Reader4 {
    /**
     * @param buf Destination buffer
     * @param n   Number of characters to read
     * @return    The number of actual characters read
     */
    private Queue<Character> queue;
    
    public Solution() {
        queue = new LinkedList<>();
    }
    
    public int read(char[] buf, int n) {
        if(n == 0)
            return 0;
        
        int count = 0;
        while(count < n) {
            char elem = getOrFillQueue();
            if(elem == 0)
                return count;
            
            buf[count++] = elem;
        }
        
        return count;
    }
    
    public char getOrFillQueue() {
        if(queue.peek() != null)
            return queue.poll();
        
        char[] buf = new char[4];
        if(read4(buf) == 0)
            return 0;
        
        // there is no charStream
        // Arrays.stream(buf).forEach(x -> queue.offer(x));
        
        for(char elem: buf)
            queue.offer(elem);
        
        return queue.poll();
        
    }
}

File: algorithms/python/436_Read_N_Characters_Given_Read4_Call_Multiple_times.py
import queue
# The read4 API is already defined for you.
# def read4(buf: List[str]) -> int:

class Solution:
    def __init__(self):
        self.queue = queue.Queue()
    
    def read(self, buf: List[str], n: int) -> int:
        if n == 0:
            return 0
        
        count = 0
        while count < n:
            elem = self.getOrFillQueue()
            if not elem:
                break
            
            buf[count] = elem
            count += 1
        
        return count
    
    def getOrFillQueue(self):
        if not self.queue.empty():
            return self.queue.get()
        
        buf = ['','','','']
        count = read4(buf)
        
        if count == 0:
            return None
        
        for elem in buf:
            self.queue.put(elem)
        
        return self.queue.get()
